hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -533,7 +533,7 @@ def first(self):
         None
         """
         # FIXME: this should generate short-circuiting IR when that is possible
-        return hl.rbind(self, lambda x: hl.or_missing(hl.len(x) > 0, x[0]))
+        return hl.fold(lambda acc, elt: hl.coalesce(acc, elt), hl.missing(self.dtype.element_type), self)
```

Comment:
is fold short-circuiting? we should nix the comment if so

Reply:
we need a `foldWithEarlyTruncation` exposed